### PR TITLE
COZMO-6362 Add support for running actions in parallel

### PR DESCRIPTION
**Pull request by MarkWesleyAtAnki**
Adding in_parallel parameter for all actions, this tells engine to run in parallel (and not clear current actions), and also prevents the SDK from raising an exception if there were pending actions in that case.
OLED face image is now an sent as a parallel action, with previous face actions automatically aborted each time
Aborted actions are immediately not considered in-progress (no longer have to wait a frame for engine to return the cancelled message before starting another action)

### CODE REVIEW

File: src/cozmo/robot.py [hunk 1]
```diff
@@ -1069,10 +1126,13 @@ def play_anim(self, name, loop_count=1):
...
         action = self.animation_factory(name, loop_count,
                 conn=self.conn, robot=self, dispatch_parent=self)
-        self._action_dispatcher._send_single_action(action)
+        self._action_dispatcher._send_single_action(action,
+                                                    in_parallel=in_parallel,
+                                                    num_retries=num_retries)
         return action
 
-    def play_anim_trigger(self, trigger, loop_count=1):
+    def play_anim_trigger(self, trigger, loop_count=1,in_parallel=False,
```
Comment (gwatts) [inline src/cozmo/robot.py]: need a space after the comma for `loop_count=1,`

File: src/cozmo/robot.py [hunk 2]
```diff
@@ -1131,9 +1199,32 @@ def display_oled_face_image(self, screen_data, duration_ms):
...
+        Raises:
+            :class:`cozmo.exceptions.RobotBusy` if another action is already
+                running and in_parallel==False
         '''
-        msg = _clad_to_engine_iface.DisplayFaceImage(faceData=screen_data, duration_ms=duration_ms)
-        self.conn.send_msg(msg)
+
+        # We never want 2 face image actions active at once, so clear current
+        # face image action (if one is running)
+        if (self._current_face_image_action is not None) and \
```
Comment (gwatts) [inline src/cozmo/robot.py]: backslash is unnecessary

File: src/cozmo/action.py [hunk 3]
```diff
@@ -91,6 +95,8 @@ def __init__(self, *, conn, robot, **kw):
         #: :class:`~cozmo.robot.Robot`: Th robot instance executing the action.
         self.robot = robot
 
+        self._action_id = None
+
```
Comment (gwatts) [inline src/cozmo/action.py]: any reason for the extra blank line here?

File: src/cozmo/action.py [hunk 4]
```diff
@@ -289,20 +302,35 @@ def _get_next_action_id(self):
             self.__class__._next_action_id += 1
         return next_action_id
 
-    def _send_single_action(self, action, position=0, num_retries=0):
+    @property
+    def _has_in_progress_actions(self):
```
Comment (gwatts) [inline src/cozmo/action.py]: would this be useful as a public property?

File: src/cozmo/action.py [hunk 5]
```diff
@@ -342,15 +370,55 @@ def _recv_msg_robot_completed_action(self, evt, *, msg):
         action_id = msg.idTag
         is_sdk_action = self._is_sdk_action_id(action_id)
         action = self._in_progress.get(action_id)
+        was_aborted = False
+        if action is None:
+            action = self._aborting.get(action_id)
+            was_aborted = (action is not None)
```
Comment (gwatts) [inline src/cozmo/action.py]: don't need the parentheses

File: src/cozmo/action.py [hunk 6]
```diff
@@ -289,20 +302,35 @@ def _get_next_action_id(self):
...
         action_id = self._get_next_action_id()
         action.robot = self.robot
+        action._action_id = action_id
 
-        if len(self._in_progress) > 0:
+        if self._has_in_progress_actions and not in_parallel:
+            # Note - it doesn't matter if previous action was started as parallel,
+            # starting any subsequent action with in_parallel==False will cancel
+            # any previous actions, so we throw an exception here and require that
+            # the client explictely cancel or wait on earlier actions
```
Comment (gwatts) [inline src/cozmo/action.py]: typo in the comment, but i appreciate it being there; preemptively answered a question ;-)

**Comment by MarkWesleyAtAnki** [comment]
I added an example program to show the parallel action usage, includes examples of edge-cases like conflicting actions (if 2 actions try to use 1 animation "track" (e.g. both try and drive the wheels) then the 2nd one returns a failure to indicate that. Also shows how each action can be independently aborted.

File: examples/tutorials/06_actions/01_parallel_actions.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    robot.set_lift_height(0.0).wait_for_completed()
+
+    cozmo.logger.info("Now, move the lift and head back up, in parallel "
+                      "- we no longer have to wait for the 1st action to complete "
+                      "before starting the 2nd one because the 2nd action is in_parallel")
+    action1 = robot.set_head_angle(cozmo.robot.MAX_HEAD_ANGLE)
+    action2 = robot.set_lift_height(1.0, in_parallel=True)
+    action1.wait_for_completed()
+    action2.wait_for_completed()
+    cozmo.logger.info("action1 = %s" % action1)
```
Comment (gwatts) [inline examples/tutorials/06_actions/01_parallel_actions.py]: can/should use a comma instead of `%` for log messages

**Comment by MarkWesleyAtAnki** [comment]
Merged via clean squashed branch - https://github.com/anki/cozmo-python-sdk/pull/64

**Pull request closed without merging**